drivers/i3c/i3c_npcx.c

```diff
@@ -0,0 +1,2055 @@
+#endif
+};
+
+struct npcx_i3c_data {
+	struct i3c_driver_data common;       /* Common i3c driver data */
+	struct k_mutex lock_mutex;           /* Mutex of i3c controller */
+	struct k_sem sync_sem;               /* Semaphore used for synchronization */
+	struct k_sem ibi_lock_sem;           /* Semaphore used for ibi */
+	enum npcx_i3c_oper_state oper_state; /* controller operation state */
+	struct {
+		/* I3C open drain clock frequency in Hz. */
+		uint32_t i3c_od_scl_hz;
+	} clocks;
+
+#ifdef CONFIG_I3C_USE_IBI
+	struct {
+		/* List of addresses used in the MIBIRULES register. */
+		uint8_t addr[5];
+
+		/* Number of valid addresses in MIBIRULES. */
+		uint8_t num_addr;
+
+		/* True if all addresses have MSB set. */
+		bool msb;
+
+		/*
+		 * True if all target devices require mandatory byte
+		 * for IBI.
+		 */
+		bool has_mandatory_byte;
+	} ibi;
+#endif
+};
+
+static void npcx_i3c_mutex_lock(const struct device *dev)
+{
+	struct npcx_i3c_data *const data = dev->data;
+
+	k_mutex_lock(&data->lock_mutex, K_FOREVER);
+}
+
+static void npcx_i3c_mutex_unlock(const struct device *dev)
+{
+	struct npcx_i3c_data *const data = dev->data;
+
+	k_mutex_unlock(&data->lock_mutex);
+}
+
+#ifdef CONFIG_I3C_NPCX_DMA
+static void i3c_ctrl_notify(const struct device *dev)
+{
+	struct npcx_i3c_data *const data = dev->data;
+
+	k_sem_give(&data->sync_sem);
+}
+
+static int i3c_ctrl_wait_completion(const struct device *dev)
+{
+	struct npcx_i3c_data *const data = dev->data;
+
+	if (k_sem_take(&data->sync_sem, I3C_TRANS_TIMEOUT) == 0) {
+		return 0;
+	}
+
+	return -ETIMEDOUT;
+}
+
+static enum npcx_i3c_oper_state get_oper_state(const struct device *dev)
+{
+	struct npcx_i3c_data *const data = dev->data;
+
+	return data->oper_state;
+}
+#endif /* CONFIG_I3C_NPCX_DMA */
+
+static void set_oper_state(const struct device *dev, enum npcx_i3c_oper_state state)
+{
+	struct npcx_i3c_data *const data = dev->data;
+
+	data->oper_state = state;
+}
+
+/*
+ * brief:  Wait for status bit done and clear the status
+ *
+ * param[in] inst  Pointer to I3C register.
+ *
+ * return  0, success
+ *         -ETIMEDOUT: check status timeout.
+ */
+static inline int npcx_i3c_status_wait_clear(struct i3c_reg *inst, uint8_t bit_offset)
+{
+	if (WAIT_FOR(IS_BIT_SET(inst->MSTATUS, bit_offset), NPCX_I3C_CHK_TIMEOUT, NULL) == false) {
+		return -ETIMEDOUT;
+	}
+
+	inst->MSTATUS = BIT(bit_offset); /* W1C */
+
+	return 0;
+}
+
+static inline uint32_t npcx_i3c_state_get(struct i3c_reg *inst)
+{
+	return GET_FIELD(inst->MSTATUS, NPCX_I3C_MSTATUS_STATE);
+}
+
+static inline void npcx_i3c_interrupt_disable(struct i3c_reg *inst, uint32_t mask)
+{
+	inst->MINTCLR = mask;
+}
+
+static inline void npcx_i3c_interrupt_enable(struct i3c_reg *inst, uint32_t mask)
+{
+	inst->MINTSET = mask;
+}
+
+static bool npcx_i3c_has_error(struct i3c_reg *inst)
+{
+	if (IS_BIT_SET(inst->MSTATUS, NPCX_I3C_MSTATUS_ERRWARN)) {
+
+		LOG_DBG("ERROR: MSTATUS 0x%08x MERRWARN 0x%08x", inst->MSTATUS, inst->MERRWARN);
```

Comment:
should this be a `LOG_ERR` or `LOG_WRN`?

Reply:
Done